app/feature/feature-edit-coinsured/src/main/kotlin/com/hedvig/android/feature/editcoinsured/ui/EditCoInsuredPresenter.kt

```diff
@@ -52,22 +57,53 @@ internal class EditCoInsuredPresenter(
         },
       )
     }
+
+    CollectEvents { event ->
+      when (event) {
+        is EditCoInsuredEvent.FetchCoInsuredPersonalInformation -> launch {
+          fetchCoInsuredPersonalInformationUseCaseProvider.provide()
+            .invoke(event.ssn)
+            .fold(
+              ifLeft = { coInsuredFromSsnError = it.message },
+              ifRight = {
+                coInsuredFromSsn = it
+                coInsuredFromSsnError = null
+              },
+            )
+        }
```

Comment:
Hmm this we can talk about a bit too.

Right now I have concerns.
1. If the member clicks this button many times, it will launch many of these coroutines here, running the query potentially many times, all in parallel. And the one which will be set in the end is just the last one which finishes, not even the last click necessarily.
2. If the member clicks the button and sends this event, there is nothing right now indicating to them that there is something loading at this moment, that we're trying to fetch something.
3. If this fails to fetch the name + last name, one approach is to show an error and try again, or we could just decide that it's fine to fail here and just let them fill those fields themselves manually, so that we do not make this step impossible to go through if for example the place where we fetch this info from is down and the query keeps failing forever. (We've had that issue with danish autocomplete once, where the service was down and having a fallback was crucial for the flow to not be completely blocked)

Very conveniently, this https://youtu.be/q9p4ewk-9E4?t=1346 video at that specific time + ~1 minute forward very specifically explains this scenario and how you can use compose state to trigger a LaunchedEffect which will actually do the work for you instead.

That solves the first 2 issues:
1. If the button is pressed again with a new input for example, then the LaunchedEffect will *cancel* the old thing it was running, and start the new one, so we don't have parallel things going on at the same time
2. We can set a loading state (not the loading of the entire screen, a loading state specifically for this query, which can power when the button in the UI shows as loading) before the query is done, and set it back to false when it's done loading.

The 3rd, to fallback in case of an error is a decision we may or may not make, we can see what makes the most sense.

Reply:
Ah found the place where we do this exact thing ourselves too, over here https://github.com/HedvigInsurance/android/blob/03cecb9b5c957d0f15084f67cbe0e49d7e102d7d/app/feature/feature-forever/src/main/kotlin/com/hedvig/android/feature/forever/ForeverViewModel.kt#L47

Reply:
Thanks, fixing